api/spellcheck.py

```diff
@@ -1,59 +1,98 @@
-        """
+        '''
+        The suggestions returned by Azure Bing Spell Check.
+        '''
 
     def __repr__(self) -> str:
         return str(self.__dict__)
 
+
+def get_key():
+    key = os.environ.get('BING_API_KEY')
+    return key
+
+
 def check(content: str) -> list[Mistake]:
-    """
+    '''
     Takes a string and returns an array of mistake objects representing
-    the errorrs in that string.
-    """
-    
-    raw_output = subprocess.run(
-        ["aspell", "-a"],
-        text=True,
-        input=content,
-        stdout=subprocess.PIPE
-    ).stdout
-    
+    the errors in that string.
+    '''
+    time.sleep(0.33334)  # Limits us to our rate on MS Azure
+    api_key = get_key()  # get key from Key Vault or ENV variables
+
+    # initalise variables to format response from Spellcheck API
     mistakes = []
-    
-    for line in raw_output.splitlines()[1:]:
-        if len(line) == 0:
-            pass
-        elif line[0] == "&":
-            word, number_of_suggestions, location, *suggestions = line.split()[1:]
-            mistakes.append(Mistake(word, int(location.rstrip(":")), suggestions))
-        elif line[0] == "#":
-            word, location = line.split()[1:]
-            mistakes.append(Mistake(word, int(location.rstrip(":")), []))
-
-    return mistakes
+    mistakes_position = []  # position of the mistakes
+    object_mistakes = []
+    suggestions = []
+    sugg = []
+    position = []  # used for for loop
+    count = 0  # used for indexing
+
+    # Set query parameters (required for the POST request to the spell check endpoint)
+    params = {
+        'mkt': 'en-US',
+        'mode': 'proof',
+        'preContextText': '',
+        'postContextText': '',
+    }
+
+    data = {'text': content}  # format json for API
+
+    # Set required headers
+    headers = {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        'Ocp-Apim-Subscription-Key': api_key,
+    }
+
+    # Send request & parse as JSON
+    response = requests.post(const.BING_ENDPOINT, headers=headers,
+                             params=params, data=data)
+    json_response = response.json()
+
+    for token in json_response[const.FLAGGED_TOKENS]:    # loop through mistakes
+        if token[const.MISTAKE_TYPE] == const.UNKOWN_TOKEN:
+            mistakes.append(token[const.TOKEN_VALUE])
+            mistakes_position.append(token[const.TOKEN_RELATIVE_POSITION])
+
+        for suggestion in token[const.SUGGESTIONS]:
+            sugg.append([suggestion])
+        suggestions.append(sugg) # just putting numbers in eg [1, 2, 3, 4, ...] based on number of mistakes
+        position.append(count)
+        count += 1
+
+    for count in position:
+        # creates object mistakes
+        accident = Mistake(
+            mistakes[count],
+            mistakes_position[count],
+            suggestions[count]
+        )
+        # creates an array of mistakes to return
+        object_mistakes.append(accident)
```

Comment:
# Question:
@thompsdy @EssienThompson @ni-sauvage @nassir90 

What is going on in this section of code?

This is my understanding:
For a mistake it loops through any suggestions and appends that suggestion to sugg as a single array, then appends this array of arrays to suggestions (so suggestions would look like
``[ [ [ suggestion1 ], [ suggestion2 ], ... ] ]``
Then it appends a count to position so for the first flagged token it would be 0, then for the second flagged token it would be 1 ect. Then when the next token comes around it appends the next tokens suggestions into sugg again then appends sugg to suggestions a second time but since sugg is never reset to an empty array now suggestions will look like
``[[[suggestion1], [suggestion2] ,...], [[[suggestion1], [suggestion2], ..., [suggestion3], [suggestion4]]]``

Is the above understanding or am I missing something? if it is correct should sugg not be reset to an empty array after each flagged token?

Reply:
It should. If there is more than one mistake in content it breaks and you get the wrong suggestions due to aliasing.